File: dao-lesson/src/main/java/com/acme/recipes/data/dao/Dao.java
package com.acme.recipes.data.dao;

import io.realm.Realm;
import io.realm.RealmModel;

public abstract class Dao<T extends RealmModel> {

    protected Realm db;

    public Dao(Realm db) {
        this.db = db;
    }

    public T copyOrUpdate(T entity) {

        if(db.isInTransaction()) {
            entity = db.copyToRealmOrUpdate(entity);
            
        } else {
            try {
                db.beginTransaction();
                entity = db.copyToRealmOrUpdate(entity);
                db.commitTransaction();
            } catch (Exception e) {
                db.cancelTransaction();
                throw e;
            }
        }

        return entity;
    }
}

File: app/src/main/java/com/acme/recipes/database/dao/Dao.kt
package com.acme.recipes.database.dao

import io.realm.Realm
import io.realm.RealmModel

abstract class Dao<T : RealmModel>(protected var db: Realm) {

    fun copyOrUpdate(entity: T): T {

        val returnEntity: T

        if (db.isInTransaction) {
            returnEntity = db.copyToRealmOrUpdate(entity)

        } else {
            try {
                db.beginTransaction()
                returnEntity = db.copyToRealmOrUpdate(entity)
                db.commitTransaction()
            } catch (e: Exception) {
                db.cancelTransaction()
                throw e
            }

        }

        return returnEntity
    }
}
